test/Test_Menhir.re

```diff
@@ -0,0 +1,216 @@
+    "let x = 5 in x",
+  ),
+  parser_test(
+    "Tuple",
+    Tuple([Int(4) |> Exp.fresh, Int(5) |> Exp.fresh]) |> Exp.fresh,
+    "(4, 5)" // TODO Verify with maketerm. Should this be parens or not
+  ),
+  parser_test(
+    "Match",
+    Match(
+      Int(4) |> Exp.fresh,
+      [
+        (Int(1) |> Pat.fresh, String("hello") |> Exp.fresh),
+        (Wild |> Pat.fresh, String("world") |> Exp.fresh),
+      ],
+    )
+    |> Exp.fresh,
+    {|case 4
+       | 1 => "hello"
+       | _ => "world"
+      end|},
+  ),
+  parser_test(
+    "If",
+    If(Bool(true) |> Exp.fresh, Int(8) |> Exp.fresh, Int(6) |> Exp.fresh)
+    |> Exp.fresh,
+    "if true then 8 else 6",
+  ),
+  parser_test(
+    "Deferred Ap",
+    DeferredAp(Var("x") |> Exp.fresh, [Deferral(InAp) |> Exp.fresh])
+    |> Exp.fresh,
+    "x(_)",
+  ),
+  parser_test(
+    "Cons",
+    Cons(Int(1) |> Exp.fresh, ListLit([]) |> Exp.fresh) |> Exp.fresh,
+    "1 :: []",
+  ),
+  parser_test(
+    "ListLit",
+    ListLit([
+      Int(1) |> Exp.fresh,
+      Int(2) |> Exp.fresh,
+      Int(3) |> Exp.fresh,
+    ])
+    |> Exp.fresh,
+    "[1, 2, 3]",
+  ),
+  menhir_only_test(
+    "Constructor",
+    Constructor("A", Unknown(Internal) |> Typ.fresh) |> Exp.fresh,
+    "A:Unknown Internal" // This is source incompatible with make_term which does not find the type. We also don't allow expression casts.
+  ),
+  parser_test(
+    "Type Alias",
+    TyAlias(Var("x") |> TPat.fresh, Int |> Typ.fresh, Int(1) |> Exp.fresh)
+    |> Exp.fresh,
+    "type x = Int in 1",
+  ),
+  parser_test(
+    "Test",
+    Test(
+      BinOp(Int(Equals), Int(3) |> Exp.fresh, Int(3) |> Exp.fresh)
+      |> Exp.fresh,
+    )
+    |> Exp.fresh,
+    "test 3 == 3 end",
+  ),
+  parser_test(
+    "Filter",
+    Filter(
+      Filter({act: (Eval, All), pat: Int(3) |> Exp.fresh}),
+      Int(3) |> Exp.fresh,
+    )
+    |> Exp.fresh,
+    "eval 3 in 3" // TODO Use other filter commands
+  ),
+  parser_test(
+    "List Concat",
+    ListConcat(
+      ListLit([Int(1) |> Exp.fresh, Int(2) |> Exp.fresh]) |> Exp.fresh,
+      ListLit([Int(3) |> Exp.fresh, Int(4) |> Exp.fresh]) |> Exp.fresh,
+    )
+    |> Exp.fresh,
+    "[1, 2] @ [3, 4]",
+  ),
+  parser_test(
+    "Integer Ops",
+    BinOp(
+      Int(GreaterThanOrEqual),
+      BinOp(
+        Int(Minus),
+        BinOp(
+          Int(Plus),
+          UnOp(Int(Minus), Int(1) |> Exp.fresh) |> Exp.fresh,
+          Int(2) |> Exp.fresh,
+        )
+        |> Exp.fresh,
+        BinOp(
+          Int(Divide),
+          Int(3) |> Exp.fresh,
+          BinOp(
+            Int(Times),
+            Int(4) |> Exp.fresh,
+            BinOp(Int(Power), Int(5) |> Exp.fresh, Int(6) |> Exp.fresh)
+            |> Exp.fresh,
+          )
+          |> Exp.fresh,
+        )
+        |> Exp.fresh,
+      )
+      |> Exp.fresh,
+      Int(8) |> Exp.fresh,
+    )
+    |> Exp.fresh,
+    "-1 + 2 - 3 / 4 * 5 ** 6 >= 8" // TODO Add the remaining operators. < is also currently broken
```

Comment:
This is currently failing due to precedence  and we can't use `<` for the reason mentioned above.